src/common/imageio_heif.c

```diff
@@ -0,0 +1,478 @@
+  int num_images = heif_context_get_number_of_top_level_images(ctx);
+  if (num_images == 0) {
+        dt_print(DT_DEBUG_IMAGEIO,
+             "No images found in HEIF file [%s]\n",
+             filename);
+    ret = DT_IMAGEIO_FILE_CORRUPTED;
+    goto out;
+  }
+
+  // We can only process a single image
+  err = heif_context_get_primary_image_handle(ctx, &handle);
+  if (err.code != 0) {
+    dt_print(DT_DEBUG_IMAGEIO,
+             "Failed to read primary image from HEIF file [%s]\n",
+             filename);
+    ret = DT_IMAGEIO_FILE_CORRUPTED;
+    goto out;
+  }
+
+  // Darktable only supports LITTLE_ENDIAN systems, so RRGGBB_LE should be fine
+  err = heif_decode_image(handle, &heif_img, heif_colorspace_RGB, heif_chroma_interleaved_RRGGBB_LE, NULL);
+  if (err.code != 0) {
+    dt_print(DT_DEBUG_IMAGEIO,
+             "Failed to decode HEIF file [%s]\n",
+             filename);
+    ret = DT_IMAGEIO_FILE_CORRUPTED;
+    goto out;
+  }
+
+  int rowbytes = 0;
+  const uint8_t* data = heif_image_get_plane_readonly(heif_img, heif_channel_interleaved, &rowbytes);
+  const size_t width = heif_image_handle_get_width(handle);
+  const size_t height = heif_image_handle_get_height(handle);
+
+  /* Initialize cached image buffer */
+  img->width = width;
+  img->height = height;
+
+  img->buf_dsc.channels = 4;
+  img->buf_dsc.datatype = TYPE_FLOAT;
+  img->buf_dsc.cst = iop_cs_rgb;
+
+  float *mipbuf = (float *)dt_mipmap_cache_alloc(mbuf, img);
+  if (mipbuf == NULL) {
+    dt_print(DT_DEBUG_IMAGEIO,
+             "Failed to allocate mipmap buffer for HEIF image [%s]\n",
+             filename);
+    ret = DT_IMAGEIO_CACHE_FULL;
+    goto out;
+  }
+
+  img->buf_dsc.filters = 0u;
+  img->flags &= ~DT_IMAGE_RAW;
+  img->flags &= ~DT_IMAGE_S_RAW;
+
+  // Get pixel value bit depth (not storage bit depth)
+  int bit_depth = heif_image_get_bits_per_pixel_range(heif_img, heif_channel_interleaved);
+
+  dt_print(DT_DEBUG_IMAGEIO,
+             "Bit depth: '%d' for HEIF image [%s]\n",
+             bit_depth,
+             filename);
+
+  /* This can be LDR or HDR, it depends on the ICC profile. But if bit_depth <= 8 it must be LDR. */
+  if(bit_depth > 8) {
+    img->flags |= DT_IMAGE_HDR;
+  } else {
+    img->flags &= ~DT_IMAGE_HDR;
+  }
+
+  float max_channel_f = (float)((1 << bit_depth) - 1);
+
+  const uint8_t *const restrict in = (const uint8_t *)data;
+
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(mipbuf, width, height, in, rowbytes, max_channel_f) \
+  schedule(simd:static) \
+  collapse(2)
+#endif
+  for (size_t y = 0; y < height; y++)
+  {
+    for (size_t x = 0; x < width; x++)
+    {
+        uint16_t *in_pixel = (uint16_t *)&in[(y * rowbytes) + (3 * sizeof(uint16_t) * x)];
+        float *out_pixel = &mipbuf[(size_t)4 * ((y * width) + x)];
+
+        /* max_channel_f is 1023.0f for 10bit */
+        out_pixel[0] = ((float)in_pixel[0]) * (1.0f / max_channel_f);
```

Comment:
I'm not sure if this remapping breaks color profiles.
It works well for Apple iPhone HEIC files (embedded ICC profile) but fails for Canon PQ Rec.2020 HEIF files (looks too dark in Darktable). But maybe Canon has done something wrong :smile:

Reply:
I've occasionally noticed some weirdness with parsing and setting input color spaces on import in general in the past... For the Canon PQ case, does it help if you set the input profile manually after import (change to something like sRGB, than change back to PQ)?

Reply:
@kmilos No, this has no effect. I've checked out how GIMP handle HEIF because the Canon HEIF files looks "okay" in GIMP. But they ignore the transfer function for PQ in NCLX and assume sRGB: https://github.com/GNOME/gimp/blob/master/plug-ins/common/file-heif.c#L780

If I open a TIFF in Gimp, convert to PQ 2020 ICC profile (generated by darktable by lcms), save as HEIF with embedded profile and open the HEIF again, the image looks to bright. I assume there is something wrong in GIMP or libheif because the same workflow works as expected for other image formats.
At this point I'm unable to produce a HEIF file with correct PQ 2020 profiling to test this with darktable.

It may also be possible that the Canon HEIF files are just not correct. Canon claims they use PQ 2020 and they write this into the NCLX block, but maybe the data is in reality sRGB or AdobeRGB.

Reply:
So does it look correct if you set sRGB or AdobeRGB input profile manually then?

Reply:
Yes, with sRGB (or AdobeRGB) input profile in Darktable all Canon HIFs looks "good", same is true for GIMP (as they use the Rec 2020 primaries but sRGB transfer curve).

Reply:
Someone in #darktable at IRC pointed me to this: https://github.com/darktable-org/darktable/issues/3880

So this problem is well known in Darktable and happen with AVIF, too.

Reply:
And if you push exposure by 6.6439 EV it is also ok, this checks out?

Reply:
No,  too much. +1.26 EV is okay but to much contrast. If you want to test the file yourself, I use some HIF files found on the internet. Tested with this one: https://www.dropbox.com/s/c19o9i9tkh2nmh4/IMG_0675.HIF?dl=0

Reply:
Is there anything I should do here? I think this get fixed with #3880 and is out of scope here?

Reply:
Indeed. I guess Canon PQ HIF files could be investigated in a separate issue if need be.